File: Baekjoon/Class3/21.1654 랜선 자르기.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split(/\s+/).map(Number);

const countCutWire = (length, wires) => { // 잘린 선 개수
    return wires.reduce((acc, wire) => acc + Math.trunc(wire / length), 0);
}

const maxCutLength = (required, wires) => {
    let answer = 0;
    let left = 1, right = Math.max(...wires);
    while(left <= right) {
        const middle = Math.trunc((left + right) / 2);
        if(countCutWire(middle, wires) >= required) { // 더 길게 자를 수도 있음
            left = middle + 1;
            answer = middle;
        }
        else // 더 짧게 잘라야 함
            right = middle - 1;
    }
    return answer;
}

const [_, n, ...wires] = input;
console.log(maxCutLength(n, wires));

File: Baekjoon/Class3/21.1654 랜선 자르기.py
import sys
input = sys.stdin.readline

def count_cut_wire(length, wires): #잘린 선 개수
    return sum(wire // length for wire in wires)

def max_cut_length(required, wires):
    answer = 0
    left = 1
    right = max(wires)
    while left <= right:
        middle = (left + right) // 2
        if count_cut_wire(middle, wires) >= required: #더 길게 자를 수도 있음
            left = middle + 1
            answer = middle
        else: #더 짧게 잘라야 함
            right = middle - 1
    return answer

def main():
    k, n = map(int, input().split())
    wires = [int(input()) for _ in range(k)]
    print(max_cut_length(n, wires)) 

main()